### DISCUSSION TITLE
How to use y_func in SlidingWindow?

### DISCUSSION THREAD

**Discussion by kiview** [timeseriesAI/tsai#104] [Q&A] [answered]
As someone without good expertise in numpy, I am lacking an example of how to use `y_func` in `SlidingWindow`, as well as an explanation of what is the default behavior.

What I want to achieve is calling something like [`scipy.stats.mode`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.mode.html) on each slice of y values, in order to calculate the corresponding y for a window.

I tried to debug the behavior by writing my own function, but it seemed my custom function received the whole numpy array (not a window) as input parameter.

Also, what is the difference between [`SlidingWindow`](https://timeseriesai.github.io/tsai/data.preparation#SlidingWindow) and [`TSUnwindowedDataset`](https://timeseriesai.github.io/tsai/data.unwindowed)?

**Comment by oguiza** [comment, marked as answer]
Hi @kiview, 

I've created a quick [gist](https://gist.github.com/oguiza/49fddc29fd8a89e61872b1967d088f63) to try to explain the difference. 

Please, let me know if you have any other questions.

**Reply by kiview** [reply]
This looks exactly like what I need, I will try it out tomorrow or at the weekend, thanks a lot!
The explanation for `TSUnwindowedDataset` is also clear and helpful.

**Reply by kiview** [reply]
Works like a charm, so accepted as the correct answer. WDYT about integrating the content of the gist to the project, as part of the documentation? I could also try to contribute it as a PR 🙂

**Reply by oguiza** [reply]
I'm glad it works for you!
I think it's a good idea. Actually, I'm thinking about creating a tutorial nb on how to prepare data. This would include what's in the gist, plus a few other methods. Thanks for your offer to create a PR 🙂. I'll let you know if I need help with it.

**Reply by bellerofonte** [reply]
Hi @oguiza!

Thank you for your [clarification](https://gist.github.com/oguiza/49fddc29fd8a89e61872b1967d088f63).
I'm trying to solve binary classification problem using `TSUnwindowedDataset` your gist. But I am facing different issues when running `lr_find` or `fit_one_cycle` those are caused by loss function computation. I've asked a [question](https://github.com/timeseriesAI/tsai/discussions/423) but possibly you could answer it here. 

I as far a I understand, then using sliding window in a batch, `y` (target) also becomes a batch of slides, so I have to transform it into batch of single values (last value of the slide), so I've defined 'y_func`:
```python
def y_func(o): 
    # o.shape is (batch_size, window_size)
    return o[:,-1]
```
And when I try to use it with `TSUnwindowedDataset`:
```python
X, y = df2xy(df.drop(columns='time'), target_col='target', steps_in_rows=False, to3d=False)
dset = TSUnwindowedDataset(X, y, y_func=y_func, window_size=WINDOW, stride=1, drop_start=0, drop_end=0, seq_first=True)
splits = TimeSplitter(valid_size=TEST_FRAC, test_size=TEST_FRAC)(range_of(len(dset)))
dsets = TSUnwindowedDatasets(dset, splits=splits)
tfms = [None, TSClassification()]
batch_tfms = TSNormalize(by_sample=True, by_var=True, range=(0,1))
dls = TSDataLoaders.from_dsets(dsets.train, dsets.valid, bs=128, tfms=tfms, batch_tfms=batch_tfms, device='cpu')
clf2 = ts_learner(dls, arch=InceptionTimePlus, loss_func=CrossEntropyLossFlat(), metrics=[RocAucBinary(), Precision()], cbs=ShowGraph())
clf2.lr_find() # <----- error
```
I get
```python
# RuntimeError: expected scalar type Long but found Float
```
If I change `y_func`:
```python
def y_func(o): 
    return o[:,-1].astype(np.int32)
```
I get another error:
```python
# IndexError: Target 1 is out of bounds.
```
If I change loss function to binary cross entropy `BCELossFlat` which seems more suitable for binary classification, I get another error:
```python
# RuntimeError: all elements of input should be between 0 and 1
```
despite my target is **either 0 or 1** and no matter which normalization I use.

What do I do wrong here?

**Reply by oguiza** [reply]
Hi @bellerofonte, sorry for the late reply. 
Could you please provide a reproducible code snippet I can run to check the issue? 
It'd be good to know the shape and type of X and y that you pass to the TSUnwindowedDataset.

**Reply by bellerofonte** [reply]
Hi @oguiza! Sorry for my late answer too.
I've made [Colab notebook](https://colab.research.google.com/drive/1jmKVVPWCeCDP2WhbJ0xmqVguO5ZPkjmg?usp=sharing) to play with. The dataset is loaded directly from at G-drive.

**Reply by oguiza** [reply]
Hi @bellerofonte, 
I've taken a quick at your code and it seems to me the issue is that you are treating the problem as a classification task when you are dealing with floats (that indicate a regression task). So should use a regression loss func and regression metrics.

**Reply by bellerofonte** [reply]
@oguiza thank you for your suggestion, but I have to clarify: do you mean `my target is float` or the `complete DS is floats`?
If `target` then there as an explicit conversion to `np.int32` in `y_func`:
```python
def y_func(o): 
    res = o[:,-1].astype(np.int32)
    return res
```
and I've add following explicit conversion before right after DS gets loaded (notebook updated):
```python
df[target_col] = df[target_col].astype(np.int32)
```
but this also has no effect

**Reply by oguiza** [reply]
Please, take a look at this [gist](https://colab.research.google.com/gist/oguiza/33c569a4660ce19ee7049396a5e9f06d/test-tsai.ipynb). There were a few things that needed to be modified in your code.

**Reply by bellerofonte** [reply]
@oguiza thanks a lot! I hope this critical step:
```python
dls.c = 2 # This is required at this stage as the TSUnwindowed dataset doesn't recognize the number of classes automatically.
```
will be mentioned in the docs one day...

... and I ran your gist "as is" and got an error at `lr_find` method:
```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-22-16f17dbd6e20>](https://localhost:8080/#) in <module>()
      1 clf2 = ts_learner(dls, arch=TST, loss_func=CrossEntropyLossFlat(), metrics=[RocAucBinary(), Precision()], cbs=ShowGraph())
----> 2 clf2.lr_find()

1 frames
[/usr/local/lib/python3.7/dist-packages/fastai/callback/schedule.py](https://localhost:8080/#) in valley(lrs, losses, num_it)
    225     idx = max_start + int(sections) + int(sections/2)
    226 
--> 227     return float(lrs[idx]), (float(lrs[idx]), losses[idx])
    228 
    229 # Cell

IndexError: index 0 is out of bounds for dimension 0 with size 0
```

**Reply by oguiza** [reply]
Sorry about that, @bellerofonte!
I've created a new [gist](https://colab.research.google.com/gist/oguiza/49c4bd9886ebdd3142473c3407b3116b/test-tsai.ipynb) based on some new functionality I've added into tsai to demonstrate a different approach I think works better, and might replace TSUnwindowedDataset in the near future. Please, take a look at it and let me know what you think.

**Reply by bellerofonte** [reply]
Hi @oguiza, I took a look at your gist and I have to say that imho this code looks much more compact and clear.
I've changed your new code to make it solve classification problem without any issues.

**Reply by oguiza** [reply]
> Hi @oguiza, I took a look at your gist and I have to say that imho this code looks much more compact and clear. I've changed your new code to make it solve classification problem without any issues.

Great!. I'm glad to hear that. I still need to decide what to do with TSUnwindowedDataset. But I also find the new approach easier to use. And in most cases, it saves RAM memory as it doesn't copy the data (it uses views).

**Reply by bellerofonte** [reply]
@oguiza, I've tried to search word *TSUnwindowedDataset* here in issues and discussions. As far as I can understand, it is not a kind of popular functionality now.

**Reply by oguiza** [reply]
I don't think it's widely used. 
The main use is when you have a large dataset and don't want or can't apply a sliding window on your data, as that may greatly increase its size. 
But the new function I've created would avoid that issue if used correctly. By that I mean you should try to use views of your data instead of copies. You can do that if you pass slices or one integer to x_vars and y_vars. If you pass an array or a list a copy will be created, and you might have an OOM error if your data is large.
I've been using this approach for a few months now and it seems to work very well on large datasets.

**Reply by bellerofonte** [reply]
yep, this approach is pretty logic. imho `TSUnwindowedDataset` should be marked as deprecated in favor of `apply_sliding_window`
